docs/DataProviders.md

```diff
@@ -650,4 +586,263 @@ const UserProfile = ({ userId }) => {
+import { LoaderFunctionArgs, json } from "@remix-run/node";
+import cloudinary from "cloudinary";
+
+export const loader = ({ request }: LoaderFunctionArgs) => {
+  cloudinary.v2.config({
+    api_key: process.env.CLOUDINARY_API_KEY,
+    api_secret: process.env.CLOUDINARY_API_SECRET,
+    cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
+    secure: false,
+  });
+
+  const timestamp = Math.round(new Date().getTime() / 1000);
+
+  const signature = cloudinary.v2.utils.api_sign_request(
+    {
+      timestamp: timestamp,
+    },
+    process.env.CLOUDINARY_API_SECRET as string
+  );
+
+  return json({
+    timestamp,
+    signature,
+    cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
+    api_key: process.env.CLOUDINARY_API_KEY,
+  });
+};
+```
+
+The following `dataProvider` extends an existing one and leverages [`withLifecycleCallbacks`](#adding-lifecycle-callbacks) to modify the `dataProvider.create()` and `dataProvider.update()` methods for the `posts` resource only with the [`beforeSave`](./withLifecycleCallbacks.md#beforesave) methods. 
+
+It works as follow:
+- it grabs the Cloudinary signature to allow the autentication in the request;
+- it creates a new `FormData` object with the file received from the form;
+- it sends this file to the Cloudinary API; 
+- it hydrates the `params.picture` object with the data sent by Cloudinary;
+- it returns the `params` object to allow the parent `dataProvider` to carry out its processes.
+
+An example is worth a thousand words:
+
+```ts
+// dataProvider.ts
+import { DataProvider, withLifecycleCallbacks } from "react-admin";
+import simpleRestProvider from "ra-data-simple-rest";
+
+type CloudinaryFile = {
+  asset_id: string;
+  secure_url: string;
+};
+
+type SignData = {
+  api_key: string;
+  timestamp: string;
+  signature: string;
+  cloud_name: string;
+};
+
+const endpoint = "http://path.to.my.api";
+
+const dataProvider = withLifecycleCallbacks(
+  simpleRestProvider(endpoint),
+  [
+    {
+      resource: "posts",
+      beforeSave: async (params: any, dataProvider: DataProvider) => {
+        const response = await fetch(
+          `${endpoint}/get-cloudinary-signature`,
+          { method: "GET" }
+          // should send headers with correct authentications
+        );
+
+        const signData: SignData = await response.json();
+
+        const url = `https://api.cloudinary.com/v1_1/${signData.cloud_name}/auto/upload`;
+
+        const formData = new FormData();
+        formData.append("file", params.picture.rawFile);
+        formData.append("api_key", signData.api_key);
+        formData.append("timestamp", signData.timestamp);
+        formData.append("signature", signData.signature);
+
+        const imageResponse = await fetch(url, {
+          method: "POST",
+          body: formData,
+        });
+
+        const image: CloudinaryFile = await imageResponse.json();
+
+        params.picture = {
+          src: image.secure_url,
+          title: image.asset_id,
+        };
```

Comment:
Avoid mutating the function parameter directly. Prefer creating a new object instead.